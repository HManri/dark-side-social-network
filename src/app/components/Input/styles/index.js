export default (theme) => ({
    'input-container': {
        boxSizing: 'border-box',
        width: '100%',
    },
    input: {
        boxSizing: 'border-box',
        backgroundColor: theme.colors.white,
        border: `1px solid ${theme.colors.border}`,
        borderRadius: '3px',
        width: '100%',
        padding: '0 10px',
        ...theme.text.big,
        lineHeight: '30px',
    },
    disabled: {
        backgroundColor: theme.colors.lightGrey,
    },
    textarea: {
        border: 'none',
        borderBottom: `1px solid ${theme.colors.border}`,
        resize: 'none',
        outline: 'none',
        ...theme.text.default,
        padding: '10px',

        '&:focus': {
            borderBottom: `1px solid ${theme.colors.lightBlack}`,
        },
    },
});
