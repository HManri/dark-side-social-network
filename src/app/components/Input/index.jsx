import React, { memo, useState, useMemo, useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { createUseStyles, useTheme } from 'react-jss';
import styles from './styles';

const useStyles = createUseStyles(styles);

const Input = memo(
    ({
        type,
        className,
        placeholder,
        value,
        onChange,
        onKeyDown,
        onKeyEnter,
        disabled,
        isTextArea,
        resetOnKeyEnter,
        controlledReset,
    }) => {
        const theme = useTheme();
        const classes = useStyles({ theme });
        const rootClassName = classnames('input', className, classes['input-container']);
        const inputClassName = classnames('input__input', classes.input, {
            [classes.disabled]: disabled,
            [classes.textarea]: !!isTextArea,
        });

        const [inputValue, setInputValue] = useState(null);

        const finalInputValue = useMemo(() => {
            if (!inputValue && inputValue !== '') return value;
            return inputValue;
        }, [value, inputValue]);

        const onChangeInput = useCallback(
            (event) => {
                const text = event.target.value;
                setInputValue(text);
                onChange && onChange(text);
            },
            [onChange],
        );

        const keyDown = useCallback(
            (evt) => {
                if (evt.keyCode === 13 && !evt.shiftKey && onKeyEnter) {
                    evt.stopPropagation();
                    evt.preventDefault();
                    onKeyEnter(inputValue);
                    resetOnKeyEnter && setInputValue('');
                }
                onKeyDown && onKeyDown(evt);
            },
            [onKeyEnter, inputValue, onKeyDown, resetOnKeyEnter],
        );

        useEffect(() => {
            if (controlledReset && inputValue && !value) {
                setInputValue('');
            }
        }, [controlledReset, inputValue, value]);

        return (
            <div className={rootClassName}>
                {!isTextArea && (
                    <input
                        type={type}
                        className={inputClassName}
                        placeholder={placeholder}
                        value={finalInputValue}
                        onChange={onChangeInput}
                        onKeyDown={keyDown}
                        disabled={disabled}
                    />
                )}
                {isTextArea && (
                    <textarea
                        className={inputClassName}
                        placeholder={placeholder}
                        value={finalInputValue}
                        onChange={onChangeInput}
                        onKeyDown={keyDown}
                    ></textarea>
                )}
            </div>
        );
    },
);

Input.defaultProps = {
    type: 'text',
    placeholder: '',
    value: '',
    disabled: false,
    isTextArea: false,
    resetOnKeyEnter: false,
    controlledReset: false,
};

Input.propTypes = {
    type: PropTypes.oneOf(['text', 'password']).isRequired,
    className: PropTypes.string,
    placeholder: PropTypes.string,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    onChange: PropTypes.func,
    onKeyDown: PropTypes.func,
    onKeyEnter: PropTypes.func,
    disabled: PropTypes.bool,
    isTextArea: PropTypes.bool,
    resetOnKeyEnter: PropTypes.bool,
    controlledReset: PropTypes.bool,
};

export default Input;
