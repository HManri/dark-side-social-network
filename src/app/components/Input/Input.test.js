import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { ThemeProvider } from 'react-jss';
import theme from '../../theme';
import Input from '.';

describe('Input', () => {
    test('should render input correctly', () => {
        const input = render(
            <ThemeProvider theme={theme}>
                <Input className="input" placeholder="Input" value="foo" />
            </ThemeProvider>,
        );

        expect(input).toMatchSnapshot();
    });

    test('should render textarea correctly', () => {
        const textarea = render(
            <ThemeProvider theme={theme}>
                <Input className="textarea" placeholder="textarea" isTextArea={true} value="foo" />
            </ThemeProvider>,
        );

        expect(textarea).toMatchSnapshot();
    });

    test('should render an input', () => {
        const { container } = render(
            <ThemeProvider theme={theme}>
                <Input className="input" placeholder="Input" value="foo" />
            </ThemeProvider>,
        );
        expect(container.firstChild.getAttribute('class')).toContain('input');
        expect(container.querySelector('input')).not.toBe(null);
        expect(container.querySelector('input').getAttribute('placeholder')).toBe('Input');
        expect(container.querySelector('input').value).toBe('foo');
    });

    test('should change its value', () => {
        const { container } = render(
            <ThemeProvider theme={theme}>
                <Input />
            </ThemeProvider>,
        );
        const htmlInput = container.querySelector('input');
        fireEvent.change(htmlInput, { target: { value: 'foo' } });
        expect(htmlInput.value).toBe('foo');
    });

    test('should be disabled', () => {
        const { container } = render(
            <ThemeProvider theme={theme}>
                <Input value="foo" disabled />
            </ThemeProvider>,
        );
        const htmlInput = container.querySelector('input');
        expect(htmlInput.value).toBe('foo');
        expect(htmlInput.hasAttribute('disabled')).toBe(true);
    });
});
