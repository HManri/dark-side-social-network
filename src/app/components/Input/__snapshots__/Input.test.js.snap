// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render input correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="input input input-container-0-2-1"
      >
        <input
          class="input__input input-0-2-2"
          placeholder="Input"
          type="text"
          value="foo"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="input input input-container-0-2-1"
    >
      <input
        class="input__input input-0-2-2"
        placeholder="Input"
        type="text"
        value="foo"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Input should render textarea correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="input textarea input-container-0-2-1"
      >
        <textarea
          class="input__input input-0-2-2 textarea-0-2-4"
          placeholder="textarea"
        >
          foo
        </textarea>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="input textarea input-container-0-2-1"
    >
      <textarea
        class="input__input input-0-2-2 textarea-0-2-4"
        placeholder="textarea"
      >
        foo
      </textarea>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
